### ISSUE TITLE
performance degrade on cupy.concatenate

### ISSUE THREAD

**Issue by uchida** [cupy/cupy#419]
With upgrading cupy version from 0.x.y to 1.0.2, we found a performance degrade on `cupy.concatenate` in our code.

the sample code is following:

```python
import sys
import cupy as xp
from cupy import cuda

cnt = 1000

n, w, h = [int(a) for a in sys.argv[1:]]
xs = xp.random.rand(n, h, w)
stream = cuda.Stream.null
start = stream.record()
for _ in range(cnt):
    xp.concatenate(xs, axis=0)
end = stream.record()
end.synchronize()

elapsed_ms = xp.cuda.get_elapsed_time(start, end) / cnt
print("{}x{}x{}: {}".format(n, h, w, elapsed_ms))
```

using chainer v1.24.0 as cupy v0.x.y and using GeForce GTX 1080 on this benchmark.

```
[cupy v0.x.y]$ python bench.py 4 1024 4096
4x1024x4096: 2.912348876953125
[cupy v1.0.2]$ python bench.py 4 1024 4096
4x1024x4096: 4.6742685546875
```

nvprof result on cupy v0.x.y(chainer v1.24.0):

```
======== Profiling result:
Time(%)      Time     Calls       Avg       Min       Max  Name
 99.69%  1.88645s      1000  1.8865ms  1.8811ms  1.8921ms  cupy_concatenate
  0.13%  2.3898ms         1  2.3898ms  2.3898ms  2.3898ms  generate_seed_pseudo(unsigned __int64, unsigned __int64, unsigned __int64, curandOrdering, curandStateXORWOW*, unsigned int*)
  0.10%  1.8539ms      3000     617ns     512ns  1.0880us  [CUDA memcpy HtoD]
  0.06%  1.1253ms         1  1.1253ms  1.1253ms  1.1253ms  cupy_random_1_minus_x
  0.03%  583.46us         1  583.46us  583.46us  583.46us  void gen_sequenced<curandStateXORWOW, double, int, __operator_&__(double curand_uniform_double_noargs<curandStateXORWOW>(curandStateXORWOW*, int))>(curandStateXORWOW*, double*, unsigned long, unsigned long, int)

======== API calls:
Time(%)      Time     Calls       Avg       Min       Max  Name
 75.17%  2.14539s      4007  535.41us     266ns  124.75ms  cudaFree
 23.47%  669.81ms      4003  167.33us  5.4030us  2.5973ms  cudaMalloc
  0.78%  22.217ms      3000  7.4050us  5.4240us  148.40us  cudaMemcpy
  0.33%  9.2933ms      1001  9.2840us  8.8430us  17.208us  cuLaunchKernel
  0.17%  4.9292ms      6007     820ns     367ns  194.71us  cudaGetDevice
  0.04%  1.1526ms         2  576.32us  575.70us  576.94us  cuDeviceTotalMem
  0.02%  693.54us       178  3.8960us     329ns  137.46us  cuDeviceGetAttribute
  0.01%  294.73us         1  294.73us  294.73us  294.73us  cudaGetDeviceProperties
  0.01%  171.43us         2  85.716us  54.733us  116.70us  cuModuleLoadData
  0.00%  78.193us         2  39.096us  37.696us  40.497us  cuDeviceGetName
  0.00%  28.008us         2  14.004us  8.5700us  19.438us  cudaLaunch
  0.00%  9.5140us         2  4.7570us     778ns  8.7360us  cudaEventDestroy
  0.00%  7.9140us         2  3.9570us  2.8170us  5.0970us  cudaEventCreateWithFlags
  0.00%  5.9650us         2  2.9820us  2.7300us  3.2350us  cudaEventRecord
  0.00%  3.9920us         4     998ns     405ns  2.5450us  cuDeviceGetCount
  0.00%  3.3270us         1  3.3270us  3.3270us  3.3270us  cudaThreadSynchronize
  0.00%  2.6960us         4     674ns     344ns  1.2650us  cuDeviceGet
  0.00%  2.3160us         4     579ns     221ns     996ns  cudaDeviceGetAttribute
  0.00%  1.8510us         1  1.8510us  1.8510us  1.8510us  cudaEventSynchronize
  0.00%  1.8000us        11     163ns     116ns     332ns  cudaSetupArgument
  0.00%  1.5510us         1  1.5510us  1.5510us  1.5510us  cudaEventElapsedTime
  0.00%  1.2860us         2     643ns     573ns     713ns  cuModuleGetFunction
  0.00%  1.2580us         4     314ns     117ns     702ns  cudaGetLastError
  0.00%  1.2530us         2     626ns     257ns     996ns  cudaConfigureCall
  0.00%  1.1020us         1  1.1020us  1.1020us  1.1020us  cuInit
  0.00%     864ns         1     864ns     864ns     864ns  cuDriverGetVersion
```

nvprof result on cupy v1.0.2:
```
======== Profiling result:
Time(%)      Time     Calls       Avg       Min       Max  Name
 99.82%  3.64150s      1000  3.6415ms  3.6210ms  3.6540ms  cupy_concatenate
  0.08%  2.8531ms         1  2.8531ms  2.8531ms  2.8531ms  generate_seed_pseudo(unsigned __int64, unsigned __int64, unsigned __int64, curandOrdering, curandStateXORWOW*, unsigned int*)
  0.05%  1.8555ms      3000     618ns     512ns  1.1200us  [CUDA memcpy HtoD]
  0.03%  1.1241ms         1  1.1241ms  1.1241ms  1.1241ms  cupy_random_1_minus_x
  0.02%  554.72us         1  554.72us  554.72us  554.72us  void gen_sequenced<curandStateXORWOW, double, int, __operator_&__(double curand_uniform_double_noargs<curandStateXORWOW>(curandStateXORWOW*, int))>(curandStateXORWOW*, double*, unsigned long, unsigned long, int)

======== API calls:
Time(%)      Time     Calls       Avg       Min       Max  Name
 84.44%  3.91567s      4007  977.21us     512ns  131.15ms  cudaFree
 14.72%  682.67ms      4003  170.54us  5.4150us  11.147ms  cudaMalloc
  0.48%  22.376ms      3000  7.4580us  5.3250us  211.11us  cudaMemcpy
  0.20%  9.2167ms      1001  9.2070us  8.7470us  16.115us  cuLaunchKernel
  0.09%  4.3477ms      6007     723ns     343ns  5.5520us  cudaGetDevice
  0.03%  1.2255ms         2  612.75us  581.70us  643.80us  cuDeviceTotalMem
  0.02%  828.53us       178  4.6540us     333ns  169.51us  cuDeviceGetAttribute
  0.01%  355.26us         1  355.26us  355.26us  355.26us  cudaGetDeviceProperties
  0.00%  170.95us         2  85.475us  60.237us  110.71us  cuModuleLoadData
  0.00%  102.25us         2  51.123us  40.931us  61.316us  cuDeviceGetName
  0.00%  24.602us         2  12.301us  9.3720us  15.230us  cudaLaunch
  0.00%  11.405us         2  5.7020us     892ns  10.513us  cudaEventDestroy
  0.00%  7.5860us         2  3.7930us  2.8410us  4.7450us  cudaEventCreateWithFlags
  0.00%  5.8310us         4  1.4570us     467ns  3.9630us  cuDeviceGetCount
  0.00%  5.8000us         2  2.9000us  2.6960us  3.1040us  cudaEventRecord
  0.00%  3.9750us         4     993ns     449ns  1.9440us  cuDeviceGet
  0.00%  3.6130us         1  3.6130us  3.6130us  3.6130us  cudaThreadSynchronize
  0.00%  2.2790us         4     569ns     246ns     946ns  cudaDeviceGetAttribute
  0.00%  1.9320us        11     175ns     116ns     378ns  cudaSetupArgument
  0.00%  1.8740us         1  1.8740us  1.8740us  1.8740us  cudaEventSynchronize
  0.00%  1.5310us         4     382ns     114ns     958ns  cudaGetLastError
  0.00%  1.4200us         1  1.4200us  1.4200us  1.4200us  cudaEventElapsedTime
  0.00%  1.2300us         2     615ns     228ns  1.0020us  cudaConfigureCall
  0.00%  1.1090us         1  1.1090us  1.1090us  1.1090us  cuInit
  0.00%  1.1070us         2     553ns     483ns     624ns  cuModuleGetFunction
  0.00%  1.0290us         1  1.0290us  1.0290us  1.0290us  cuDriverGetVersion
```

Average elapsed time on `cupy_concatenate` kernel and `cudaFree` API call are degraded on v1.0.2.
As far as I confirmed, concatenate kernel has no changes since chainer v1.21.0, so I couldn't catch the root-cause.

In addition, for this sample, I found `cupy.concatenate` on chainer v1.20.0 (cupy_copy with skips) does the best performance in any versions of cupy (so currently we replace concatenate with older one in our code).
On chainer v1.21.0, some optimizations were done on concatenate, but it seems some implicit assumption on these optimizations, probably total size or shape, are broken in this code.

**Comment by okuta** [comment]
Current `cupy.concatenate` has two copy method. 
CuPy chooses which to use depending on the number of sequences to combine.
We should change that [threshold](https://github.com/cupy/cupy/blob/5f68c474fa1e61504bcbaa16136516a3a8cc1119/cupy/core/core.pyx#L2409).

If you have a time. Cloud you send PR?

**Comment by uchida** [comment]
Giving new threshold seems good and I could make a time to send PR, but I'm afraid that optimal threshold may differ between cupy version (0.x, 1.x, 2.x) since concatenate benchmark results differ between cupy version (0.x, 1.x, 2.x) even they use same concatenate kernel/function.
I'm suspicious about cupy memory layout affects this, but I'm not sure about it.

**Comment by okuta** [comment]
I think that optimum is difficult. Let's aim for better things.

**Comment by uchida** [comment]
OK, I'll work on it.

**Comment by kmaehashi** [comment]
Already resolved in #452
